plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/StandardFacetInstallDelegate.java

```diff
@@ -75,6 +75,7 @@ private AppEngineRuntimeInstallJob(IFacetedProject facetedProject) {
       super(Messages.getString("appengine.install.runtime.to.project", // $NON-NLS$
           facetedProject.getProject().getName()));
       this.facetedProject = facetedProject;
+      setRule(facetedProject.getProject());
```

Comment:
> The first ConvertJob is the one that installs the JSDT facet, and it does commitChanges(). (I don't think it tries to acquire a workspace lock, but I think it'll be blocked on the lock for modifying project facets anyway.)

`FacetedProjectWorkingCopy#commitChanges()` calls `FacetedProject#mergeChanges()` which does [acquire the workspace lock](http://git.eclipse.org/c/webtools-common/webtools.common.git/tree/plugins/org.eclipse.wst.common.project.facet.core/src/org/eclipse/wst/common/project/facet/core/internal/FacetedProject.java?h=master#n1109)

> I think there should be a way to apply a scheduling rule to some portion of code of a Job, rather than applying it to the entire duration of the job? If we grab a lock after waitUntilJsdtIsFixedFacet(), at least we can get away with this deadlock.

Yeah, we should be able to wrap `AppEngineStandardFacet#installAppEngineRuntime()`.  In fact, maybe it's best to put these inside `installAppEngineRuntime()` and `installAppEngineFacet()`.